include/NAS2D/Filesystem.h

```diff
@@ -50,16 +57,64 @@ class Filesystem
 
 	void toggleVerbose() const;
 
+    void setWorkingDirectory(const std::filesystem::path& p);
+    std::filesystem::path getWorkingDirectory();
+
 private:
-	Filesystem(const Filesystem&) = delete;
-	Filesystem& operator= (const Filesystem&) = delete;
 
-	bool closeFile(void *file) const;
-	const char* getLastPhysfsError() const;
+    std::filesystem::path getExePath() const;
 
-private:
-	std::string			mDataPath;			/**< Data path string. This will typically be 'data/'. */
-	mutable bool		mVerbose;			/**< Displays lots of messages when true. Otherwise only critical messages are displayed. */
+    bool isInit() const;
+    bool exists(const std::filesystem::path& p) const;
+
+    void forEachFileInFolder(const std::filesystem::path& folderpath, const std::string& validExtensionList = std::string{}, const std::function<void(const std::filesystem::path&)>& callback = [](const std::filesystem::path& p) { (void)p; }, bool recursive = false) const;
+    void forEachFileInFolder(const std::filesystem::path& folderpath, const std::string& validExtensionList = std::string{}, const std::function<void(const std::filesystem::path&)>& callback = [](const std::filesystem::path& p) { (void)p; }, bool recursive = false);
+
+    template<typename DirectoryIteratorType>
+    void forEachFileInFolder(const std::filesystem::path& folderpath, const std::string& validExtensionList = std::string{}, const std::function<void(const std::filesystem::path&)>& callback = [](const std::filesystem::path& p) { (void)p; }, bool recursive = false) const;
+
+    bool readBufferFromFile(std::vector<unsigned char>& out_buffer, const std::string& filePath) const;
+    bool readBufferFromFile(std::string& out_buffer, const std::string& filePath) const;
+
+    bool writeBufferToFile(void* buffer, std::size_t size, const std::string& filePath) const;
+    bool writeBufferToFile(const std::string& buffer, const std::string& filePath) const;
+
+
+    mutable std::vector<std::filesystem::path> mSearchPath{};
```

Comment:
I'm unfamiliar with this style of in-place initialization -- was this C++11 or later? I'm not at all opposed, just curious. I usually use '='. Is there a difference?

Reply:
This was a standards conformance change that was added to VS2017 late last year, I can't find changelog that mentions the exact version of 15.x it was. It's called default member initialization and prevents out-of-order initialization in c'tor initializer lists errors (most novice and beginner modern c++ programmers don't know about the rule that members initialize in order of declaration in the header not in order of initializer list in the implementation) and makes it so you only have to initialize the non-default members in the initializer list or are able to get rid of default c'tors entirely by either with the use of "= default" (good) or not declaring them at all and using the compiler generated c'tor (better) instead. A very useful change to the language that I've started to love.